File: shared_resources/src/main/java/com/thewizrd/shared_resources/weatherdata/WeatherManager.java
package com.thewizrd.shared_resources.weatherdata;

import android.location.Location;

import androidx.annotation.WorkerThread;

import com.thewizrd.shared_resources.BuildConfig;
import com.thewizrd.shared_resources.controls.LocationQueryViewModel;
import com.thewizrd.shared_resources.locationdata.LocationData;
import com.thewizrd.shared_resources.locationdata.LocationProviderImpl;
import com.thewizrd.shared_resources.tasks.AsyncTask;
import com.thewizrd.shared_resources.tasks.CallableEx;
import com.thewizrd.shared_resources.utils.Settings;
import com.thewizrd.shared_resources.utils.WeatherException;
import com.thewizrd.shared_resources.utils.WeatherUtils;
import com.thewizrd.shared_resources.weatherdata.here.HEREWeatherProvider;
import com.thewizrd.shared_resources.weatherdata.metno.MetnoWeatherProvider;
import com.thewizrd.shared_resources.weatherdata.nws.NWSWeatherProvider;
import com.thewizrd.shared_resources.weatherdata.openweather.OpenWeatherMapProvider;
import com.thewizrd.shared_resources.weatherdata.weatherunlocked.WeatherUnlockedProvider;
import com.thewizrd.shared_resources.weatherdata.weatheryahoo.YahooWeatherProvider;

import java.util.Collection;

// Wrapper class for supported Weather Providers
public final class WeatherManager implements WeatherProviderImplInterface {
    private static WeatherManager instance;
    private static WeatherProviderImpl weatherProvider;

    // Prevent instance from being created outside of this class
    private WeatherManager() {
        updateAPI();
    }

    public static synchronized WeatherManager getInstance() {
        if (instance == null)
            instance = new WeatherManager();

        return instance;
    }

    public void updateAPI() {
        String API = Settings.getAPI();
        weatherProvider = getProvider(API);
    }

    // Static Methods
    public static WeatherProviderImpl getProvider(String API) {
        WeatherProviderImpl providerImpl = null;

        switch (API) {
            case WeatherAPI.YAHOO:
                providerImpl = new YahooWeatherProvider();
                break;
            case WeatherAPI.HERE:
                providerImpl = new HEREWeatherProvider();
                break;
            case WeatherAPI.OPENWEATHERMAP:
                providerImpl = new OpenWeatherMapProvider();
                break;
            case WeatherAPI.METNO:
                providerImpl = new MetnoWeatherProvider();
                break;
            case WeatherAPI.NWS:
                providerImpl = new NWSWeatherProvider();
                break;
            case WeatherAPI.WEATHERUNLOCKED:
                providerImpl = new WeatherUnlockedProvider();
                break;
            default:
                if (!BuildConfig.DEBUG) {
                    providerImpl = new WeatherUnlockedProvider();
                }
                break;
        }

        if (providerImpl == null)
            throw new IllegalArgumentException("Argument API: Invalid API name! This API is not supported");

        return providerImpl;
    }

    public static boolean isKeyRequired(String API) {
        WeatherProviderImpl provider = getProvider(API);
        return provider.isKeyRequired();
    }

    public static boolean isKeyValid(final String key, final String API) throws WeatherException {
        final WeatherProviderImpl provider = getProvider(API);
        return AsyncTask.await(new CallableEx<Boolean, WeatherException>() {
            @Override
            public Boolean call() throws WeatherException {
                return provider.isKeyValid(key);
            }
        });
    }

    // Provider dependent methods
    @Override
    public String getWeatherAPI() {
        return weatherProvider.getWeatherAPI();
    }

    @Override
    public boolean isKeyRequired() {
        return weatherProvider.isKeyRequired();
    }

    @Override
    public boolean supportsWeatherLocale() {
        return weatherProvider.supportsWeatherLocale();
    }

    @Override
    public boolean supportsAlerts() {
        return weatherProvider.supportsAlerts();
    }

    @Override
    public boolean needsExternalAlertData() {
        return weatherProvider.needsExternalAlertData();
    }

    @Override
    public void updateLocationData(LocationData location) {
        weatherProvider.updateLocationData(location);
    }

    @Override
    public String updateLocationQuery(final Weather weather) {
        return weatherProvider.updateLocationQuery(weather);
    }

    @Override
    public String updateLocationQuery(final LocationData location) {
        return weatherProvider.updateLocationQuery(location);
    }

    @WorkerThread
    @Override
    public Collection<LocationQueryViewModel> getLocations(final String ac_query) throws WeatherException {
        return weatherProvider.getLocations(ac_query);
    }

    @WorkerThread
    public LocationQueryViewModel getLocation(final Location location) throws WeatherException {
        return weatherProvider.getLocation(new WeatherUtils.Coordinate(location));
    }

    @WorkerThread
    @Override
    public LocationQueryViewModel getLocation(final WeatherUtils.Coordinate coordinate) throws WeatherException {
        return weatherProvider.getLocation(coordinate);
    }

    @WorkerThread
    @Override
    public Weather getWeather(final String location_query, final String country_code) throws WeatherException {
        return weatherProvider.getWeather(location_query, country_code);
    }

    @WorkerThread
    @Override
    public Weather getWeather(final LocationData location) throws WeatherException {
        return weatherProvider.getWeather(location);
    }

    @WorkerThread
    @Override
    public Collection<WeatherAlert> getAlerts(final LocationData location) {
        return weatherProvider.getAlerts(location);
    }

    @Override
    public String localeToLangCode(String iso, String name) {
        return weatherProvider.localeToLangCode(iso, name);
    }

    @Override
    public String getWeatherIcon(String icon) {
        return weatherProvider.getWeatherIcon(icon);
    }

    @Override
    public String getWeatherIcon(boolean isNight, String icon) {
        return weatherProvider.getWeatherIcon(isNight, icon);
    }

    @Override
    public String getWeatherCondition(String icon) {
        return weatherProvider.getWeatherCondition(icon);
    }

    @Override
    public boolean isKeyValid(String key) throws WeatherException {
        return weatherProvider.isKeyValid(key);
    }

    @Override
    public String getAPIKey() {
        return weatherProvider.getAPIKey();
    }

    @Override
    public boolean isNight(Weather weather) {
        return weatherProvider.isNight(weather);
    }

    @Override
    public LocationProviderImpl getLocationProvider() {
        return weatherProvider.getLocationProvider();
    }
}

File: shared_resources/src/main/java/com/thewizrd/shared_resources/weatherdata/WeatherManager.kt
package com.thewizrd.shared_resources.weatherdata

import android.location.Location
import androidx.annotation.WorkerThread
import com.thewizrd.shared_resources.BuildConfig
import com.thewizrd.shared_resources.SimpleLibrary
import com.thewizrd.shared_resources.controls.LocationQueryViewModel
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.locationdata.LocationProviderImpl
import com.thewizrd.shared_resources.utils.Coordinate
import com.thewizrd.shared_resources.utils.WeatherException
import com.thewizrd.shared_resources.weatherdata.here.HEREWeatherProvider
import com.thewizrd.shared_resources.weatherdata.metno.MetnoWeatherProvider
import com.thewizrd.shared_resources.weatherdata.nws.NWSWeatherProvider
import com.thewizrd.shared_resources.weatherdata.openweather.OpenWeatherMapProvider
import com.thewizrd.shared_resources.weatherdata.weatherunlocked.WeatherUnlockedProvider
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class WeatherManager private constructor() : WeatherProviderImplInterface {
    companion object {
        private var sInstance: WeatherManager? = null
        private var sWeatherProvider: WeatherProviderImpl? = null

        @JvmStatic
        val instance: WeatherManager
            get() {
                if (sInstance == null) {
                    sInstance = WeatherManager()
                }

                return sInstance!!
            }

        @JvmStatic
        fun getProvider(API: String?): WeatherProviderImpl {
            var providerImpl: WeatherProviderImpl? = null

            when (API) {
                WeatherAPI.HERE -> providerImpl = HEREWeatherProvider()
                WeatherAPI.OPENWEATHERMAP -> providerImpl = OpenWeatherMapProvider()
                WeatherAPI.METNO -> providerImpl = MetnoWeatherProvider()
                WeatherAPI.NWS -> providerImpl = NWSWeatherProvider()
                WeatherAPI.WEATHERUNLOCKED -> providerImpl = WeatherUnlockedProvider()
                else -> {
                    if (!BuildConfig.DEBUG) {
                        providerImpl = WeatherUnlockedProvider()
                    }
                }
            }

            requireNotNull(providerImpl) { "Argument API: Invalid API name! This API is not supported" }

            return providerImpl
        }

        @JvmStatic
        fun isKeyRequired(API: String): Boolean {
            val provider = getProvider(API)
            return provider.isKeyRequired()
        }

        @Throws(WeatherException::class)
        @JvmStatic
        suspend fun isKeyValid(key: String?, API: String): Boolean {
            val provider = getProvider(API)
            return withContext(Dispatchers.Default) {
                provider.isKeyValid(key)
            }
        }
    }

    init {
        updateAPI()
    }

    fun updateAPI() {
        val settingsMgr = SimpleLibrary.getInstance().app.settingsManager
        val API = settingsMgr.getAPI()
        sWeatherProvider = getProvider(API)
    }

    // Provider dependent methods
    override fun getWeatherAPI(): String {
        return sWeatherProvider!!.getWeatherAPI()
    }

    override fun isKeyRequired(): Boolean {
        return sWeatherProvider!!.isKeyRequired()
    }

    override fun supportsWeatherLocale(): Boolean {
        return sWeatherProvider!!.supportsWeatherLocale()
    }

    override fun supportsAlerts(): Boolean {
        return sWeatherProvider!!.supportsAlerts()
    }

    override fun needsExternalAlertData(): Boolean {
        return sWeatherProvider!!.needsExternalAlertData()
    }

    override fun getHourlyForecastInterval(): Int {
        return sWeatherProvider!!.getHourlyForecastInterval()
    }

    override suspend fun updateLocationData(location: LocationData) {
        sWeatherProvider!!.updateLocationData(location)
    }

    override fun updateLocationQuery(weather: Weather): String {
        return sWeatherProvider!!.updateLocationQuery(weather)
    }

    override fun updateLocationQuery(location: LocationData): String {
        return sWeatherProvider!!.updateLocationQuery(location)
    }

    override fun getLocationProvider(): LocationProviderImpl {
        return sWeatherProvider!!.getLocationProvider()
    }

    @WorkerThread
    @Throws(WeatherException::class)
    override suspend fun getLocations(ac_query: String?): Collection<LocationQueryViewModel> {
        return sWeatherProvider!!.getLocations(ac_query)
    }

    @WorkerThread
    @Throws(WeatherException::class)
    override suspend fun getLocation(location: Location): LocationQueryViewModel? {
        return sWeatherProvider!!.getLocation(Coordinate(location))
    }

    @WorkerThread
    @Throws(WeatherException::class)
    override suspend fun getLocation(coordinate: Coordinate): LocationQueryViewModel? {
        return sWeatherProvider!!.getLocation(coordinate)
    }

    @WorkerThread
    @Throws(WeatherException::class)
    override suspend fun getWeather(location_query: String, country_code: String): Weather {
        return sWeatherProvider!!.getWeather(location_query, country_code)
    }

    @WorkerThread
    @Throws(WeatherException::class)
    override suspend fun getWeather(location: LocationData?): Weather {
        return sWeatherProvider!!.getWeather(location)
    }

    @WorkerThread
    override suspend fun getAlerts(location: LocationData): Collection<WeatherAlert>? {
        return sWeatherProvider!!.getAlerts(location)
    }

    override fun localeToLangCode(iso: String, name: String): String {
        return sWeatherProvider!!.localeToLangCode(iso, name)
    }

    override fun getWeatherIcon(icon: String?): String {
        return sWeatherProvider!!.getWeatherIcon(icon)
    }

    override fun getWeatherIcon(isNight: Boolean, icon: String?): String {
        return sWeatherProvider!!.getWeatherIcon(isNight, icon)
    }

    override fun getWeatherCondition(icon: String?): String {
        return sWeatherProvider!!.getWeatherCondition(icon)
    }

    @Throws(WeatherException::class)
    override suspend fun isKeyValid(key: String?): Boolean {
        return sWeatherProvider!!.isKeyValid(key)
    }

    override fun getAPIKey(): String? {
        return sWeatherProvider!!.getAPIKey()
    }

    override fun isNight(weather: Weather): Boolean {
        return sWeatherProvider!!.isNight(weather)
    }
}
